### ISSUE TITLE
Component : CTA - Screen reader - When the modal dialog is displayed, screen reader still identifies the content from the parent page.

### ISSUE THREAD

**Issue by Praveen-sr** [carbon-design-system/ibm-dotcom-library#2727]
## Detailed description

Issue :- Video Modal - When the modal dialog is displayed, screen reader still identifies the content from the parent page.

Expected result :- 
When the modal dialog is displayed, the parent page content should be hidden from the screen reader.

**Comment by guilhermelMoraes** [comment]
@Praveen-sr would you please clarify how to reproduce this bug? Maybe a step-by-step, what's happening and what's expected

**Comment by Praveen-sr** [comment]
@guilhermelMoraes 

1. Open storybook and navigate to CTA
2. In knobs section, edit/select "Type" as "Video"
3. Click on the video link
4. Modal opens

Issue :- For screen readers, once the modal is open, while navigating using up/down arrows it still reads the parent page details (The video link used to trigger the modal)

**Comment by caiobozato** [comment]
@Praveen-sr I've checked and tested the code and the screen reader is not reading the CTA behind the Modal. What happens is that the video `title` and it's `duration` are appearing inside an `aria-label` attribute:

![image.png](https://images.zenhubusercontent.com/5eed1fec732800bd837801c2/cc8a09d8-1cf8-4e4d-a928-f533b78c8e8a)

That is, the screen reader is reading the `aria-label` property in the video player.

**Comment by caiobozato** [comment]
@Praveen-sr I forgot to mention that I removed the `aria-label` mentioned above using the browser inspector and the screen reader stopped "saying" the text we had in CTA, that is, definitely the screen reader is not reading the CTA behind the modal.

**Comment by Praveen-sr** [comment]
@caiobozato okay, i'll ask Ram to re test and confirm. Can you give the link where to verify this ?
Thanks!

**Comment by caiobozato** [comment]
@Praveen-sr I've used the link for the latest version live of the library: https://ibmdotcom-react.mybluemix.net/?path=/story/components-cta--default

But you can use any of the links (canary, experimental, etc.) Just remember to inspect and look for the `aria-label` under the vídeo `div` like the screenshot I sent above.

**Comment by Praveen-sr** [comment]
@caiobozato Ram re tested this on Canary env,
The focus is trapped within the dialog. Screen reader identifies the Dialog Role as well.
Now the issue is There is no label for the dialog.

**Comment by caiobozato** [comment]
@Praveen-sr Thanks for clarifying that.

About the label, you mean the modal should have `aria-label`? If yes, which text should be added there?

Also, which HTML element of modal should receive this `label`? 

I'm asking that because part of this component comes from Carbon, depending on your requirements, the approach to fix that will be different/

**Comment by caiobozato** [comment]
@Praveen-sr Any updates?

**Comment by Praveen-sr** [comment]
@caiobozato The role=”dialog” container can be given label via either aria-labelledby attribute that references the title of the dialog, or an aria-label if there is no visible title. When there is no visible title, generally the value for aria label can be taken from the control which triggers the modal If it is concise and meaningful.

**Comment by asudoh** [comment]
@Praveen-sr Thank you for the info, one quick question; Does the problem described here happen at https://react.carbondesignsystem.com/iframe.html?id=modal--default? If so, how do you think it should work?

**Comment by Praveen-sr** [comment]
@asudoh Your asking us to verify the aria-label on the above link ?

**Comment by asudoh** [comment]
Thank you for asking @Praveen-sr - Not `aria-label` technically, but the screen reader behavior in above link. Depending on how well Carbon core `<ComposedModal>` is working with the screen reader, we can determine how technically we can solve the problem.

**Comment by Praveen-sr** [comment]
@asudoh Okay, got you. Thanks. We'll check on the link and update you.

**Comment by Praveen-sr** [comment]
@asudoh @caiobozato Ram tested above example in FF with JAWS:
In the given example, When we activate “Launch Composed Modal” button, We get a modal dialog for which the dialog role is identified and the focus also trapped within the dialog. However this dialog does not have any label.

**Comment by asudoh** [comment]
Thanks @Praveen-sr - When you say "this dialog does not have any label", did you mean you expected the screen reader speaks about modal title or modal description but didn't? Or something else?

**Comment by Praveen-sr** [comment]
@asudoh Screen reader is expected to announce the title i.e. Accessible name for the dialog. As I explained earlier, The Accessible name is Derived from the aria-labelledby ID which refers to the main heading in the dialog. If there is no visible text which can be associated using aria-labeledby, aria-label can be used instead. Request you to go through this aria 1.1 spec example link for reference.
https://www.w3.org/TR/wai-aria-practices-1.1/examples/dialog-modal/dialog.html

**Comment by caiobozato** [comment]
@Praveen-sr Thanks for the info. @asudoh should this update be done in Carbon modal? Because I noticed the `LightboxMediaViewer` component uses it as a dependency.

**Comment by asudoh** [comment]
@caiobozato I see the content is defined in [our code](https://github.com/carbon-design-system/ibm-dotcom-library/blob/v1.8.0/packages/react/src/components/LightboxMediaViewer/LightboxMediaViewer.js#L79-L85) and thus I think we need to define `aria-labeledby` by our own. Would you try something like this? https://github.com/carbon-design-system/ibm-dotcom-library/pull/3180/files

**Comment by caiobozato** [comment]
@Praveen-sr First, thanks for sharing the info and `aria` docs. Checking the docs I saw the `aria-labelledby` it's always on the same `div` as `role="dialog"` I have a question: is this mandatory? I mean must they always be in the same `element`? 

------

@asudoh If the answer from Praveen is yes, I've made some tests and need your help:

I tried to apply an `aria-labelledby` attribute in our `LightboxMediaViewer`, in a couple code parts:

![image.png](https://images.zenhubusercontent.com/5eed1fec732800bd837801c2/563d476b-5f70-4db6-9c81-5f3765c9f340)


But on rendering, the `aria-labelledby` attribute is applied to elements that don't have the `role="dialog"` attribute.

![image.png](https://images.zenhubusercontent.com/5eed1fec732800bd837801c2/86ca47cd-020b-46da-b7a7-80a448cdfcfa)

Tried on `ExpressiveModal.js` as well: 

![image.png](https://images.zenhubusercontent.com/5eed1fec732800bd837801c2/b56c8034-d136-427e-8084-e0102fe86954)

![image.png](https://images.zenhubusercontent.com/5eed1fec732800bd837801c2/17493b10-c63a-4dc0-9bb2-fa1dce48d590)

I think the `<div class="bx--modal-container" role="dialog">` is part of `Carbon` code and not `IBM.com library`.

Any thoughts?

**Comment by asudoh** [comment]
@caiobozato The code example I shared before provides a way to change the DOM that Carbon core renders. Does it make sense?

**Comment by caiobozato** [comment]
@asudoh Yes, that makes sense now. Thanks Akira!

**Comment by caiobozato** [comment]
@Praveen-sr Could you and you team please test again? https://ibmdotcom-react.s3.us-south.cloud-object-storage.appdomain.cloud/deploy-previews/3275/index.html?path=/story/overview-getting-started--page

**Comment by Praveen-sr** [comment]
@caiobozato @RobertaJHahn Ram re tested this on the link provided above, now the Screen reader focus is trapped within the dialog in arrow key navigation and the label is also announced. Working as expected.

**Comment by caiobozato** [comment]
@Praveen-sr Thank you very much!

**Comment by caiobozato** [comment]
@RobertaJHahn by my conversation with @asudoh I still need to mock a value on testing file for `LightboxMediaViewer` component. 

I'm moving back to review, if there's no problem.
